src/flamenco/runtime/fd_executor.c

```diff
@@ -1184,89 +1184,84 @@ fd_instr_stack_pop( fd_exec_txn_ctx_t *       txn_ctx,
-                  fd_instr_info_t *   instr ) {
+  fd_instr_info_t *   instr ) {
   FD_RUNTIME_TXN_SPAD_FRAME_BEGIN( txn_ctx->spad, txn_ctx ) {
-  FD_SCRATCH_SCOPE_BEGIN {
-    fd_exec_instr_ctx_t * parent = NULL;
-    if( txn_ctx->instr_stack_sz ) {
-      parent = &txn_ctx->instr_stack[ txn_ctx->instr_stack_sz - 1 ];
-    }
-
-    int err = fd_instr_stack_push( txn_ctx, instr );
-    if( FD_UNLIKELY( err ) ) {
-      FD_TXN_ERR_FOR_LOG_INSTR( txn_ctx, err, txn_ctx->instr_err_idx );
-      return err;
-    }
-
-    fd_exec_instr_ctx_t * ctx = &txn_ctx->instr_stack[ txn_ctx->instr_stack_sz - 1 ];
-    *ctx = (fd_exec_instr_ctx_t) {
-      .instr     = instr,
-      .txn_ctx   = txn_ctx,
-      .epoch_ctx = txn_ctx->epoch_ctx,
-      .slot_ctx  = txn_ctx->slot_ctx,
-      .acc_mgr   = txn_ctx->acc_mgr,
-      .funk_txn  = txn_ctx->funk_txn,
-      .parent    = parent,
-      .index     = parent ? (parent->child_cnt++) : 0,
-      .depth     = parent ? (parent->depth+1    ) : 0,
-      .child_cnt = 0U,
-    };
-
-    txn_ctx->instr_trace[ txn_ctx->instr_trace_length - 1 ] = (fd_exec_instr_trace_entry_t) {
-      .instr_info = instr,
-      .stack_height = txn_ctx->instr_stack_sz,
-    };
-
-    fd_exec_instr_fn_t native_prog_fn = fd_executor_lookup_native_program( &txn_ctx->borrowed_accounts[ instr->program_id ] );
-    fd_exec_txn_ctx_reset_return_data( txn_ctx );
-    int exec_result = FD_EXECUTOR_INSTR_SUCCESS;
-    if( native_prog_fn != NULL ) {
-      /* Log program invokation (internally caches program_id base58) */
-      fd_log_collector_program_invoke( ctx );
-      exec_result = native_prog_fn( ctx );
-    } else {
-      exec_result = FD_EXECUTOR_INSTR_ERR_UNSUPPORTED_PROGRAM_ID;
-    }
-
-    int stack_pop_err = fd_instr_stack_pop( txn_ctx, instr );
-    if( FD_LIKELY( exec_result == FD_EXECUTOR_INSTR_SUCCESS ) ) {
-      /* Log success */
-      fd_log_collector_program_success( ctx );
-
-      /* Only report the stack pop error on success */
-      if( FD_UNLIKELY( stack_pop_err ) ) {
-        FD_TXN_ERR_FOR_LOG_INSTR( txn_ctx, stack_pop_err, txn_ctx->instr_err_idx );
-        return stack_pop_err;
+    FD_SCRATCH_SCOPE_BEGIN {
+      fd_exec_instr_ctx_t * parent = NULL;
+      if( txn_ctx->instr_stack_sz ) {
+        parent = &txn_ctx->instr_stack[ txn_ctx->instr_stack_sz - 1 ];
       }
-    } else {
-      /* if txn_ctx->exec_err is not set, it indicates an instruction error */
-      if( !txn_ctx->exec_err ) {
-        FD_TXN_ERR_FOR_LOG_INSTR( txn_ctx, exec_result, txn_ctx->instr_err_idx );
+
+      // https://github.com/anza-xyz/agave/blob/b1acfbde5bd78488641e7dca3cafc07bef48b87b/program-runtime/src/invoke_context.rs#L251
+      //
+      // the get_last_program_key() throws the error that causes the UNSUPPORTED_PROGRAM_ID
+      int err = ( instr->acct_cnt == 0 ) ? FD_EXECUTOR_INSTR_ERR_UNSUPPORTED_PROGRAM_ID : fd_instr_stack_push( txn_ctx, instr );
+      if( FD_UNLIKELY( err ) ) {
```

Comment:
From what I can tell, `InstructionContext.get_last_program_key` uses the _transaction's_ program indicies list. This is given to the instruction context at instruction preperation time, but is ultimately populated when the accounts for the _transaction_ are loaded: https://github.com/anza-xyz/agave/blob/e063c19025d65870a513fcc010c47685dc989d52/svm/src/account_loader.rs#L476

This uses each instruction's `program_id_index` (https://github.com/anza-xyz/agave/blob/e063c19025d65870a513fcc010c47685dc989d52/svm-transaction/src/instruction.rs#L9), _not_ the list of instruction accounts (https://github.com/anza-xyz/agave/blob/e063c19025d65870a513fcc010c47685dc989d52/svm-transaction/src/instruction.rs#L11).

So this check needs to be done at the transaction level, using each instruction's `program_id` field and checking that this corresponds to a _transaction account_.